Use connctionhelper to connect to remote Docker host.

guineveresaenger opened the pull request:
Fixes #662 
Fixes #580

t0yv0 commented on provider/go.mod:
```diff
@@ -9,15 +9,15 @@ require (
-	github.com/pulumi/pulumi/sdk/v3 v3.66.0
-	github.com/stretchr/testify v1.8.2
+	github.com/pulumi/pulumi-terraform-bridge/v3 v3.50.1
+	github.com/pulumi/pulumi/pkg/v3 v3.71.0
```
Would be best if the updates were a separate (automated) PR.

guineveresaenger replied:
I can run that first if you prefer

guineveresaenger replied:
https://github.com/pulumi/pulumi-docker/pull/669

t0yv0 commented on provider/image.go:
```diff
@@ -813,13 +813,26 @@ func configureDockerClient(configs map[string]string) (*client.Client, error) {
+		if err != nil {
+			return nil, err
+		}
+		if helper != nil {
```
I don't understand this bit here, is there ever the case where helper==nil? Oh

```It returns nil without error when no helper is registered for the scheme.```

guineveresaenger replied:
added explanation!

t0yv0 commented:
This needs tests. Thoughts on adding tests? I understand that standing up an SSH docker host is perhaps problematic for integration testing here? Can we consider factoring out the code to use unit tests, or is it possible to run ProgramTest with an SSH based docker host without actually standing it up, and observe some meaningful difference in before/after the change that we can assert on?

AaronFriel commented:
Agree with @t0yv0 on wanting to test this:

>  Can we consider factoring out the code to use unit tests,

This seems like the lightest weight approach; can we create docker contexts in a unit test (without any running Docker host) and test that the helper uses the docker context selected?

guineveresaenger commented:
I added a few unit tests that verify ssh-based behavior.

It turns out that the SSH connection helper _only_ supports ssh-based scheme, and it _always_ returns `http://docker.example.com` as the client's daemon host.

I leverage this knowledge in the tests as checking for `http://docker.example.com` as the daemon host, and added a few default checks 🤞

necaris commented:
I hope this will make it into a 4.2.5 that will be released soon?
